js/modules/k6/experimental/streams/readable_streams.go

```diff
@@ -0,0 +1,586 @@
+//
+// [ReadableStreamError]: https://streams.spec.whatwg.org/#readable-stream-error
+func (stream *ReadableStream) error(e any) {
+	// 1. Assert: stream.[[state]] is "readable".
+	if stream.state != ReadableStreamStateReadable {
+		common.Throw(stream.vu.Runtime(), newError(AssertionError, "cannot error a stream that is not readable"))
+	}
+
+	// 2. Set stream.[[state]] to "errored".
+	stream.state = ReadableStreamStateErrored
+
+	// 3. Set stream.[[storedError]] to e.
+	stream.storedError = e
+
+	// 4. Let reader be stream.[[reader]].
+	reader := stream.reader
+
+	// 5. If reader is undefined, return.
+	if reader == nil {
+		return
+	}
+
+	genericReader, ok := reader.(ReadableStreamGenericReader)
+	if !ok {
+		common.Throw(stream.vu.Runtime(), newError(AssertionError, "reader is not a ReadableStreamGenericReader"))
+	}
+
+	// 6. Reject reader.[[closedPromise]] with e.
+	promise, _, rejectFunc := genericReader.GetClosed()
+	rejectFunc(e)
+
+	// 7. Set reader.[[closedPromise]].[[PromiseIsHandled]] to true.
+	// See https://github.com/dop251/goja/issues/565
+	var (
+		err       error
+		doNothing = func(goja.Value) {}
+	)
+	_, err = promiseThen(stream.vu.Runtime(), promise, doNothing, doNothing)
+	if err != nil {
+		common.Throw(stream.vu.Runtime(), newError(RuntimeError, err.Error()))
+	}
+
+	// 8. If reader implements ReadableStreamDefaultReader,
+	defaultReader, ok := reader.(*ReadableStreamDefaultReader)
+	if ok {
+		// 8.1. Perform ! ReadableStreamDefaultReaderErrorReadRequests(reader, e).
+		defaultReader.errorReadRequests(e)
+		return
+	}
+
+	// 9. OTHERWISE, reader is a ReadableStreamBYOBReader
+	// 9.1. Assert: reader implements ReadableStreamBYOBReader.
+	common.Throw(stream.vu.Runtime(), newError(RuntimeError, "ReadableStreamBYOBReader is not supported yet"))
+}
+
+// fulfillReadRequest implements the [ReadableStreamFulfillReadRequest()] algorithm.
+//
+// [ReadableStreamFulfillReadRequest()]: https://streams.spec.whatwg.org/#readable-stream-fulfill-read-request
+func (stream *ReadableStream) fulfillReadRequest(chunk any, done bool) {
+	// 1. Assert: ! ReadableStreamHasDefaultReader(stream) is true.
+	if !stream.hasDefaultReader() {
+		common.Throw(stream.vu.Runtime(), newError(AssertionError, "stream does not have a default reader"))
+	}
+
+	// 2. Let reader be stream.[[reader]].
+	reader, ok := stream.reader.(*ReadableStreamDefaultReader)
+	if !ok {
+		common.Throw(stream.vu.Runtime(), newError(RuntimeError, "reader is not a ReadableStreamDefaultReader"))
+	}
+
+	// 3. Assert: reader.[[readRequests]] is not empty.
+	if len(reader.readRequests) == 0 {
+		common.Throw(stream.vu.Runtime(), newError(AssertionError, "reader.[[readRequests]] is empty"))
+	}
+
+	// 4. Let readRequest be reader.[[readRequests]][0].
+	readRequest := reader.readRequests[0]
+
+	// 5. Remove readRequest from reader.[[readRequests]].
+	reader.readRequests = reader.readRequests[1:]
+
+	callback := reader.vu.RegisterCallback()
```

Comment:
I am also not certain why this is used.

In all of the places where RegisterCallback is used currently it is on the event loop and it queues otehr stuff on the event loop.

none of the tests fail if I remove any of htem.

The only place where you go off the event loop is in ReadableStream#Cancel and ther you use `promises.New` :+1: 

So I am not certain why this and the other RegisterCallback exist.

Reply:
Same, same as with `closeSteps` and `errorSteps`. Thanks!